### PR TITLE
Refactor node draining racing avoid condition

### PR DESCRIPTION
**Pull request by pliurh**
Utilize k8s leader election mechanism to prevent annotating nodes
at the same time.

### CODE REVIEW

File: pkg/daemon/daemon.go
Comment (zshi-redhat) [inline pkg/daemon/daemon.go:846]: Can we run `dn.getDrainLock` in foreground and not use `done` chan? 
I assume `leaderelection.RunOrDie` is a loop w/o timeout, is it true?

File: pkg/daemon/daemon.go
Comment (adrianchiris) [inline pkg/daemon/daemon.go:809]: looking at leaderelection [docs](https://pkg.go.dev/k8s.io/client-go@v0.19.10/tools/leaderelection):

> Package leaderelection implements leader election of a set of endpoints. It uses an annotation in the endpoints object to store the record of the election state. This implementation does not guarantee that only one client is acting as a leader (a.k.a. fencing).

is it not an issue ?
how will the system behave when there is only one endpoint trying to take the lead on LeaseLock ?

File: pkg/daemon/daemon.go
Comment (adrianchiris) [inline pkg/daemon/daemon.go:822]: using this mechanism, do we still need to annotate the node with `annotDraining` ?

Comment (pliurh) [reply pkg/daemon/daemon.go:822]: That is the trick. The leader election mechanism requires the leader to keep updating the Lease object. But in our case, the node may reboot itself, then lose leadership. So I use a 2 layers lock here. The node can only start draining with 2 conditions: 1) it becomes the leader 2) no other node is draining which is indicated by the annotation.

File: pkg/daemon/daemon.go
Comment (pliurh) [reply pkg/daemon/daemon.go:846]: The `leaderelection.RunOrDie` runs forever unless the context is canceled. I didn't run it in the foreground, because I want it keeps leading until the node finishes draining. So the rest nodes can only start the election after the prior one finishes draining if there is no reboot required.

File: pkg/daemon/daemon.go
Comment (adrianchiris) [reply pkg/daemon/daemon.go:809]: OK thanks for explaining

File: pkg/daemon/daemon.go
Comment (pliurh) [reply pkg/daemon/daemon.go:820]: That is intentional. We don't want a configuration mistake to break more nodes. If users encounter such problem, they'd better do some troubleshooting to find why the node cannot come back.

File: pkg/daemon/daemon.go
Comment (pliurh) [reply pkg/daemon/daemon.go]: fixed

File: pkg/daemon/daemon.go
Comment (pliurh) [reply pkg/daemon/daemon.go]: fixed

**Review by martinkennelly — APPROVED** [review]

**Review by adrianchiris — APPROVED** [review]
I would prefer the package updates be performed in a separate commit in the PR (easier to review this way). but would not block on it.

@SchSeba i see below you were requested as reviewer. once you approve this can be merged IMO

File: pkg/daemon/daemon.go
Comment (adrianchiris) [reply pkg/daemon/daemon.go:823]: @pliurh mind mentioning this in the commit message ? so its clear in the commit message how this mechanism is used to control node draining

**Pull request merged**
